# Importing Libraries

In [2]:
#Import libraries
import pandas as pd
import numpy as np
import os

# Import Data

In [5]:
#Define path
path = r'/Users/arturassada/Desktop/Instacart Basket Analysis'

In [6]:
#import products.csv and order.csv

In [53]:
# import order.csv as df_ords
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'orders.csv'), index_col = False)

In [10]:
# import order.csv as df_ords
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'products.csv'), index_col = False)

# 4.5 Exercise operations

In [17]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [18]:
#Create dataframe
df_test = pd.DataFrame()

In [19]:
#Create a mixed-type column
df_test['mix'] = ['a', 'b', 1, True]

In [20]:
df_test


mix
0     a
1     b
2     1
3  True

In [1]:
#Check data for mixed-type data
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

NameError: name 'df_test' is not defined

In [22]:
df_test['mix'] = df_test['mix'].astype('str')

In [23]:
#Find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [26]:
#Find missing values rows
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [28]:
#Print missing values
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [29]:
#df shape
df_prods.shape

(49693, 5)

In [31]:
#Create clean data set
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [32]:
#Data Shape check after removing missing values
df_prods_clean.shape

(49677, 5)

In [34]:
#Drop NaN from dataset
df_prods.dropna(subset = ['product_name'], inplace = True)

In [35]:
#check for duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [38]:
#Check for duplicates
df_dups.shape

(5, 5)

In [39]:
#Data
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [40]:
#create new df without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [41]:
#new df shape
df_prods_clean_no_dups.shape

(49672, 5)

In [ ]:
#Exporting to CSV file
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task 4.5

In [ ]:
#1. If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now.
# ->> done above.

In [47]:
#2. Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
#Tip: Keep an eye on min and max values!
df_prods.describe()
    

product_id      aisle_id  department_id        prices
count  49677.000000  49677.000000   49677.000000  49677.000000
mean   24850.194235     67.763110      11.728687      9.993164
std    14340.588602     38.316396       5.850651    453.592708
min        1.000000      1.000000       1.000000      1.000000
25%    12433.000000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.000000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

Comments:
1. min for all columns is the same, a bit suspicious
2. price of 99999 - probbaly an issue


In [48]:
#3. Check for mixed-type data in your df_ords dataframe.

In [56]:
#Check if data has mixed data types in columns 
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)

In [68]:
#4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

#-> no mixed data found


In [69]:
#5. Run a check for missing values in your df_ords dataframe.
#In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [70]:
#Find missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [71]:
#6. Address the missing values using an appropriate method. In a markdown cell, explain why you used your method of choice.

Comments:
1. No missing values found
2. If found optrions are:
    a) ignore - not advised
    b) impute median or average
    c) remove - if relatively small amount

In [72]:
#7. Run a check for duplicate values in your df_ords data.
#In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [82]:
#Create a subset df with duplicates
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [83]:
#Check df shape
df_ords_dups.shape

(0, 7)

In [84]:
#Print duplicates
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

Comment:
1. duplicates found
2. Only few compared to saple size -> remove them
3. Duplicates could happen in transactional system or data migration or ETL

In [85]:
#remove duplicates from df_ords (main order dataframe)


In [86]:
#8. Address the duplicates using an appropriate method.
#In a markdown cell, explain why you used your method of choice.
df_ords_clean = df_ords.drop_duplicates()

Comment:
1. Chose to remove since only 5 rows

In [87]:
df_ords_clean.shape

(3421083, 7)

In [88]:
#9. Export data

In [89]:
#export clean products.csv data
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned_20220614.csv'))

In [90]:
#export products products.csv data
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_cleaned_20220614.csv'))